### PR TITLE
EVM-578: Deploy new contracts for staking and supernets

### PR DESCRIPTION
**Pull request by goran-ethernal**
# Description

This PR deploys new contracts from RFC-201 regarding validator storage rework.

New contracts on `rootchain`:
- `StakeManager` - handles staking on `root`.
- `CustomSupernetManager` - handles `whitelist`, `registering` validators, and `genesis` validators.

New contracts on `sidechain`:
- `RewardDistributor` - handles rewards distribution.
- `ValidatorSet` - basically replaces the old `ChildValidatorSet` contract.

What this PR doesn't do:
- changing the e2e tests
- replacing the old `ChildValidatorSet` (this will be replaced, through next PRs, once we start staking on root).

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [x] I have tested this code manually

### CODE REVIEW

File: command/rootchain/initcontracts/init_contracts.go [hunk 1]
```diff
@@ -263,50 +311,61 @@ func deployContracts(outputter command.OutputFormatter, client *jsonrpc.Client,
 			name:     rootERC1155PredicateName,
 			artifact: contractsapi.RootERC1155Predicate,
 		},
+		{
+			name:     stakeManagerName,
+			artifact: contractsapi.StakeManager,
+			constructorCallback: func(artifact *artifact.Artifact, cfg *polybft.RootchainConfig) ([]byte, error) {
+				encoded, err := artifact.Abi.Constructor.Inputs.Encode([]interface{}{cfg.RootNativeERC20Address})
```
Comment (Stefan-Ethernal) [inline command/rootchain/initcontracts/init_contracts.go:318]: Just curious would go bindings work with constructors? Maybe worth trying out. WDYT?

File: command/genesis/polybft_params.go [hunk 2]
```diff
@@ -289,6 +294,61 @@ func (p *genesisParams) deployContracts(totalStake *big.Int) (map[types.Address]
...
+			},
+		},
+		{
+			artifact: contractsapi.RewardDistributor,
+			address:  contracts.RewardDistributorContract,
+			constructorCallback: func(artifact *artifact.Artifact,
+				polybftConfig *polybft.PolyBFTConfig,
+				rootConfig *polybft.RootchainConfig) ([]byte, error) {
+				encoded, err := artifact.Abi.Constructor.Inputs.Encode([]interface{}{
+					contracts.NativeERC20TokenContract, // TODO @goran-ethernal - we will use a different token for rewards
```
Comment (Stefan-Ethernal) [inline command/genesis/polybft_params.go:338]: Just to add my two cents to TODO:
maybe we should consider to parameterize whether we are using native token for rewards or different one. However it is a question for product.

File: command/genesis/polybft_params.go [hunk 3]
```diff
@@ -289,6 +294,61 @@ func (p *genesisParams) deployContracts(totalStake *big.Int) (map[types.Address]
...
+						"stake": validator.Stake,
+					}
+				}
+
+				// TODO @goran-ethernal - we will remove once we change e2e tests
+				// since by RFC-201 we won't be able to start edge without bridge (root)
+				customSupernetManagerAddr := types.ZeroAddress
+				if rootConfig != nil {
+					customSupernetManagerAddr = rootConfig.CustomSupernetManagerAddress
+				}
```
Comment (Stefan-Ethernal) [inline command/genesis/polybft_params.go]: Change this and we can remove `*polybft.RootchainConfig` param from the function signature alltogether. WDYT?
```suggestion
				if polybftConfig.Bridge != nil {
					customSupernetManagerAddr = polybftConfig.Bridge.CustomSupernetManagerAddr
				}
```

**Review by Stefan-Ethernal — APPROVED** [review]
LGTM

File: command/rootchain/initcontracts/init_contracts.go (hunk 1 above)
Comment (goran-ethernal) [reply command/rootchain/initcontracts/init_contracts.go:318]: Not sure really. Let's try that through a separate PR.

File: command/genesis/polybft_params.go (hunk 2 above)
Comment (goran-ethernal) [reply command/genesis/polybft_params.go:338]: Agreed.

File: command/genesis/polybft_params.go (hunk 3 above)
Comment (goran-ethernal) [reply command/genesis/polybft_params.go]: Fix: https://github.com/0xPolygon/polygon-edge/pull/1368/commits/7cec43f8fd68498c9589fe0c4867df9890532f78

**Review by igorcrevar — APPROVED** [review]

**Pull request merged**